Store line change in script navigation history

KoBeWi opened the pull request:
Closes #16554
Closes https://github.com/godotengine/godot-proposals/issues/4989 (what a coincidence...)

Any movement with more than 10 lines apart is saved into history.
![godot windows tools 64_qixu8sT8AL](https://user-images.githubusercontent.com/2223172/181123686-b9c2c307-02dd-4e21-a7eb-624401e8f85a.gif)
The implementation is rather hacky and might be not 100% reliable, but script editors are a mess :/ At least it works.

reduz commented:
I have the feeling I implemented this at some distant time in the past..

Calinou commented:
> I have the feeling I implemented this at some distant time in the past..

You might be thinking about script tab temperature, which is a different feature that serves different needs;

reduz commented:
@Calinou No I am pretty sure I made the history go back to the proper script/line.

KoBeWi commented:
Yes, the history works this way, but the line change wasn't recorded in the history. Only script changes and in 3.x Ctrl+clicks on members (which was probably a convenient bug lol).

reduz commented:
Oh alright, in any case, up to @Paulb23 to review

Paulb23 reviewed (dismissed):
Yep definitely a slightly hacky approach, using `[set|get]_edit_state` will also change breakpoints and bookmarks etc, so currently we could end up resting these as well, which is not ideal.

Should also probably add support for scrolling around the docs as well.

KoBeWi commented:
> Should also probably add support for scrolling around the docs as well.

Docs don't have caret though.

KoBeWi commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -786,6 +794,14 @@ void ScriptTextEditor::_breakpoint_toggled(int p_row) {
+void ScriptTextEditor::_on_caret_moved() {
+	int current_line = code_editor->get_text_editor()->get_caret_line();
+	if (ABS(current_line - previous_line) > 10) {
+		emit_signal(SNAME("request_save_previous_state"), previous_line);
```
I don't think there is a way to use a signal without exposing it, is there?

Paulb23 replied:
I mean't not exposing the `[get|set]_previous_state` methods by emitting the state directly here (only navigation state).

Paulb23 commented:
For docs I was thinking along the lines of, clicking some in doc navigation such as a method in the same class. It would be nice to store that as a navigation action.

KoBeWi committed:
Store line change in script navigation history

KoBeWi commented:
Updated. I changed to use the new navigation history and simplified storing a bit. Also I added some scroll saving to editor help, but it's unreliable.

There is a bug sometimes where a position can be saved twice in a history, but I don't know what triggers it.

KoBeWi commented:
I did some more tweaks and I think this works reliably now.

akien-mga reviewed (approved):
Tested, seems to work fine.

akien-mga commented:
Thanks!

TheDying0fLight commented on editor/plugins/script_text_editor.h:
```diff
@@ -260,6 +263,9 @@ class ScriptTextEditor : public ScriptEditorBase {
 	virtual void validate() override;
 
+	Variant get_previous_state();
+	void store_previous_state();
```
I found these on master and looked for their usage, but from what I can tell these the `get` functions aren't called and `previous_state` is not used directly. So the functions and the code for them in `CodeTextEditor` seem to do nothing. Am I missing something or can these be removed?
